Fix Corefile migration logic

rajansandeep opened the pull request.

fturib commented on cluster/gce/config-default.sh:
```diff
@@ -195,12 +195,13 @@ if [[ ! -z "${NODE_ACCELERATORS}" ]]; then
 # Optional: Install cluster DNS.
-ENABLE_CLUSTER_DNS="${KUBE_ENABLE_CLUSTER_DNS:-true}"
+ENABLE_CLUSTER_CORE_DNS="${KUBE_ENABLE_CLUSTER_CORE_DNS:-true}"
+ENABLE_CLUSTER_KUBE_DNS="${KUBE_ENABLE_CLUSTER_KUBE_DNS:-false}"
```
I would propose something like:

keep : ENABLE_CLUSTER_DNS (True/False) - means it expect installation of  DNS Service (no change here)
add : CLUSTER_DNS_IS_CORE_DNS (True/False) - which define what service to install. (if the value does not exist then t is evaluated to False and then KubeDNS is installed, as before).

fturib commented on cluster/openstack-heat/kubernetes-heat/fragments/configure-salt.yaml:
```diff
@@ -53,7 +53,8 @@ write_files:
-      enable_cluster_dns: "true"
+      enable_cluster_kube_dns: "true"
+      enable_cluster_core_dns: "false"
       dns_server: 10.246.0.10
```
There is no validation of XOR on the enable flags.
I guess the last installed is winning ? is that sure ? I mean : is there some artifact not cleaned up if you install KubeDNS + CoreDNS ?

fturib commented on cluster/photon-controller/config-default.sh:
```diff
@@ -74,7 +74,8 @@ ELASTICSEARCH_LOGGING_REPLICAS=1
 # Optional: Install cluster DNS.
-ENABLE_CLUSTER_DNS="${KUBE_ENABLE_CLUSTER_DNS:-true}"
+ENABLE_CLUSTER_KUBE_DNS="${KUBE_ENABLE_CLUSTER_KUBE_DNS:-true}"
+ENABLE_CLUSTER_CORE_DNS="${KUBE_ENABLE_CLUSTER_CORE_DNS:-true}"
```
- there is a change of default logic for the user : he need to define which DNS service, therefore know both of them

fturib commented on cluster/vagrant/provision-utils.sh:
```diff
@@ -62,7 +62,8 @@ enable_cluster_ui: '$(echo "$ENABLE_CLUSTER_UI" | sed -e "s/'/''/g")'
 elasticsearch_replicas: '$(echo "$ELASTICSEARCH_LOGGING_REPLICAS" | sed -e "s/'/''/g")'
-enable_cluster_dns: '$(echo "$ENABLE_CLUSTER_DNS" | sed -e "s/'/''/g")'
+enable_cluster_kube_dns: '$(echo "$ENABLE_CLUSTER_KUBE_DNS" | sed -e "s/'/''/g")'
+enable_cluster_core_dns: '$(echo "$ENABLE_CLUSTER_CORE_DNS" | sed -e "s/'/''/g")'
```
- if we look into the code, it is heavier because need to double the tests for setting up the same parts of code.

fturib commented on cluster/libvirt-coreos/user_data_minion.yml:
```diff
@@ -20,8 +20,11 @@ coreos:
+        $( [[ "$ENABLE_CLUSTER_KUBE_DNS" == "true" ]] && echo "--cluster-domain=${DNS_DOMAIN}" ) \
+        $( [[ "$ENABLE_CLUSTER_CORE_DNS" == "true" ]] && echo "--cluster-dns=${DNS_SERVER_IP}" ) \
+        $( [[ "$ENABLE_CLUSTER_CORE_DNS" == "true" ]] && echo "--cluster-domain=${DNS_DOMAIN}" ) \
+        $( [[ "$ENABLE_CLUSTER_CORE_DNS" == "true" ]] && echo "--service-cluster-ip-range=${SERVICE_CLUSTER_IP_RANGE}" ) \
```
- if we look into the code, it is heavier because need to double the tests for setting up the same parts of code.

chrisohaver commented:
we should close this PR and move discussion to the new one #2

rajansandeep committed:
fix migration logic

modify addon test to cover the entire migration logic

chrisohaver commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -280,6 +280,7 @@ func createCoreDNSAddon(deploymentBytes, serviceBytes, configBytes []byte, clien
 	canMigrateCorefile := true
 
 	if corefileMigrationRequired {
+		//If migration is required, try and migrate the Corefile
```
We get in here for default corefiles that would require migration (e.g. when new defaults are added, or default options removed)... but we should instead be replacing those with the current default.

chrisohaver commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -289,10 +290,16 @@ func createCoreDNSAddon(deploymentBytes, serviceBytes, configBytes []byte, clien
 		}
-	} else {
+	} else if isCoreDNSCorefileEmptyOrDefault(corefile) {
+		// If the Corefile is empty or default, the latest default Corefile will be applied
```
I think the only time we get in here is when the default corefile is empty or already identical to the new default corefile (no migrations required).

chrisohaver commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -367,12 +374,17 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 }
 
+// isCoreDNSCorefileEmptyOrDefault checks if the Corefile is empty or a unmodified Corefile
+func isCoreDNSCorefileEmptyOrDefault(corefile string) bool {
```
I don't think this needs to be a separate function.

chrisohaver commented:
Regarding replacing vs migrating a default corefile... it seems a minor point since essentially the end result is the same.  
I think replacing might be safer long term because of the possibility of an errant migration migrating default users "off" the default track.  I lean toward replacing over always migrating.
However, it would be simpler code wise to just always migrate in which case we never need to check too see if it's default or not. 

What do you think?

rajansandeep commented:
> Regarding replacing vs migrating a default corefile... it seems a minor point since essentially the end result is the same.
> I think replacing might be safer long term because of the possibility of an errant migration migrating default users "off" the default track. I lean toward replacing over always migrating.
> However, it would be simpler code wise to just always migrate in which case we never need to check too see if it's default or not.
> 
> What do you think?

I think I agree that if the Corefile is default, we should just replace it rather than migrating it.
I think we can achieve that by not checking for `newdefaults`, since the default check includes any version of CoreDNS and should end up just replacing it when upgrading.

I'll add my next commit with debug statements and test results to see if it makes sense.

chrisohaver commented:
I think default Corefiles are getting migrated, but the test doesn't care

Can you add a test case where the corefile is the default, but plugins not in the same order as the default (`migration.Default` should still return true in that case - it's not order sensitive).  If the Corefile is getting replaced, the outcome would be plugins ordered in the default order.  If it's getting migrated, it would retain the altered order of plugins from the original Corefile.  The test also needs to migrate to a version where the default Corefile has changed.

rajansandeep commented:
> I think this is still an issue. For a non-default Corefile with no removed deprecations, `isCoreDNSConfigMapMigrationRequired` will return false. It will then retain the Corefile, thus no new defaults would be added if there were any.

You're right, I reverted the `newdefault` check for this and added a test case covering this.

chrisohaver commented:
Looks good.  One nit: I think that `isCoreDNSCorefileEmptyOrDefault` is too simple to be a separate function that is only used once, and considering it could be a one line function, e.g. ...

```
func isCoreDNSCorefileEmptyOrDefault(corefile string) bool {
 	return corefile == "" || migration.Default("", corefile)
 }
```

rajansandeep commented:
> Looks good. One nit: I think that `isCoreDNSCorefileEmptyOrDefault` is too simple to be a separate function that is only used once, and considering it could be a one line function, e.g. ...
> 
> ```
> func isCoreDNSCorefileEmptyOrDefault(corefile string) bool {
>  	return corefile == "" || migration.Default("", corefile)
>  }
> ```

cool! I removed it, if it looks good, I'll squash and send the PR upstream
